### ISSUE TITLE
hyper 0.13 support (std futures)

### ISSUE THREAD

**Issue by lnicola** [scottlamb/http-serve#16]

**Comment by lnicola** [comment]
Nb. `hyper 0.13` is now out.

**Comment by scottlamb** [comment]
I've started flailing at this until the compiler is happy. "cargo build" now succeeds locally with upgraded Rust, tokio, and hyper. "cargo build --features test" doesn't succeed yet, much less an actual "cargo test" run.

**Comment by scottlamb** [comment]
Most of the tests now compile and pass. I'm having trouble concatenating a body stream. In particular, this `src/chunker.rs` code:

```rust
    use futures::stream::StreamExt;

    ...

    type BodyStream = Box<dyn Stream<Item = Result<Vec<u8>, ()>> + Send + Sync>;

    // A smaller-than-chunk-size write shouldn't be flushed on write, and there's currently no Drop
    // implementation to do it either.
    #[test]
    fn small_no_flush() {
        let (mut w, body): (_, BodyStream) = BodyWriter::with_chunk_size(4);
        assert_eq!(w.write(b"1").unwrap(), 1);
        w.truncate();
        drop(w);
        assert_eq!(b"", &body.concat().wait().unwrap()[..]);
    }
```

fails to compile:

```
error: the `concat` method cannot be invoked on a trait object
   --> src/chunker.rs:131:31
    |
131 |         assert_eq!(b"", &body.concat().wait().unwrap()[..]);
    |                               ^^^^^^
    |
    = note: another candidate was found in the following trait, perhaps add a `use` for it:
            `use futures_util::stream::stream::StreamExt;`
```

[futures 0.3.1 StreamExt::concat](https://docs.rs/futures/0.3.1/futures/stream/trait.StreamExt.html#method.concat)

[futures 0.1.29 Stream::concat2](https://docs.rs/futures/0.1.29/futures/stream/trait.Stream.html#method.concat2) (which worked)

`Stream` (and thus `StreamExt`) should be implemented here both by the trait object and by its outer Box. It's like it's trying to call concat on the trait object, when I want to call it on the outer box. There's probably something dumb I'm missing. I blame jet lag.

**Comment by lnicola** [comment]
There's a https://docs.rs/hyper/0.13.1/hyper/body/fn.to_bytes.html function which might work here.

**Comment by scottlamb** [comment]
Hmm, that does work, if I alter the body stream to use `Bytes` as the data type and `BoxedError` as the error type so I can convert it to a `HttpBody`. Still curious why I can't get `concat` to work like before.

Progress. Gotta get ready for work now though.

**Comment by scottlamb** [comment]
> `Stream` (and thus `StreamExt`) should be implemented here both by the trait object and by its outer Box.

Ahh, turns out it isn't implemented by its outer box because a `dyn Stream` is not `Unpin`. It is implemented by a `Pin<Box<...>>` so I think I just have to use `Pin<_>::from(...)` on the box to get something I can call these methods on.

I'm glad to have figured this out because there are other methods on `StreamExt` with the same problem.

**Comment by scottlamb** [comment]
Patch below with an unsuccessful attempt to reintroduce InnerBody.

[innerbody-attempt.patch.txt](https://github.com/scottlamb/http-serve/files/3976972/innerbody-attempt.patch.txt)

and a key error:

```
error[E0599]: no method named `poll_next` found for type `std::pin::Pin<std::boxed::Box<dyn futures_core::stream::Stream<Item = std::result::Result<D, E>> + std::marker::Send + std::marker::Sync>>` in the current scope
   --> src/serving.rs:226:39
    |
226 |                 Pin::new_unchecked(b).poll_next(ctx)
    |                                       ^^^^^^^^^ method not found in `std::pin::Pin<std::boxed::Box<dyn futures_core::stream::Stream<Item = std::result::Result<D, E>> + std::marker::Send + std::marker::Sync>>`
```

Somehow `flatten()` is able to work with a `Pin<Box<dyn Stream<...> + ...>`. But I can't call `.poll_next()` on one. Not sure how to reconcile that.

**Comment by scottlamb** [comment]
Hmm, a variation of my code will compile and work:

```rust
enum InnerBody<D, E> {
    Once(Option<D>),
    B(Pin<Box<dyn Stream<Item = Result<D, E>> + Sync + Send>>),
}

impl<D, E> Stream for InnerBody<D, E> where D: Unpin
{
    type Item = Result<D, E>;
    fn poll_next(self: Pin<&mut Self>, ctx: &mut futures::task::Context) -> std::task::Poll<Option<Result<D, E>>> {
        match self.get_mut() {
            InnerBody::Once(ref mut o) => std::task::Poll::Ready(o.take().map(|d| Ok(d))),
            InnerBody::B(b) => {
                b.as_mut().poll_next(ctx)
            },
        }
    }
}
```

it requires the chunk type be `Unpin`. I _think_ that won't be a problem for callers. I could use `self.get_unchecked_mut()` to avoid that but I don't understand pinning well enough (yet) to know if that's sound.

**Comment by lnicola** [comment]
I think `Unpin` is just fine. Actually, using `Bytes` would probably have been good enough. Sorry I can't meaningfully review your changes otherwise, the whole pinning thing kind of eludes me..

On the topic of benchmarks, you might want to be aware of two issues with `tokio::fs`:

 - it limits the buffer size to 16 KB
 - it's much more slow than in `tokio 0.1` right now

For more details, see the linked issues in https://github.com/stephank/hyper-staticfile/issues/24.

**Comment by scottlamb** [comment]
I managed to convince myself the unsafe is sound anyway—it's pretty close to the example in the std::pin docs.

Thanks for the pointer to that issue. Since http-serve calls std::io::File methods directly, the buffer size thing isn't a problem, but the contact switches might be relevant.

**Comment by scottlamb** [comment]
Published http-serve 0.2.0 based on hyper 0.13.x and friends.
